libs/cluster/Server/Replication/ReplicaOps/ReplicaReplayTask.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Threading;
+using Garnet.common;
+using Microsoft.Extensions.Logging;
+using Tsavorite.core;
+
+namespace Garnet.cluster
+{
+    internal sealed partial class ReplicationManager : IBulkLogEntryConsumer, IDisposable
+    {
+        TsavoriteLogScanSingleIterator replayIterator = null;
+        CancellationTokenSource replicaReplayTaskCts;
+        SingleWriterMultiReaderLock activeReplay;
+
+        /// <summary>
+        /// Reset background replay iterator
+        /// </summary>
+        public void ResetReplayIterator()
+        {
+            ResetReplayCts();
+            replayIterator?.Dispose();
+            replayIterator = null;
+
+            void ResetReplayCts()
+            {
+                if (replicaReplayTaskCts == null)
+                {
+                    replicaReplayTaskCts = CancellationTokenSource.CreateLinkedTokenSource(ctsRepManager.Token);
+                }
+                else
+                {
+                    replicaReplayTaskCts.Cancel();
+                    try
+                    {
+                        activeReplay.WriteLock();
+                        replicaReplayTaskCts.Dispose();
+                        replicaReplayTaskCts = CancellationTokenSource.CreateLinkedTokenSource(ctsRepManager.Token);
+                    }
+                    finally
+                    {
+                        activeReplay.WriteUnlock();
+                    }
+                }
+            }
+        }
+
+        public void Throttle() { }
+
+        public unsafe void Consume(byte* record, int recordLength, long currentAddress, long nextAddress, bool isProtected)
+        {
+            ReplicationOffset = currentAddress;
+            var ptr = record;
+            while (ptr < record + recordLength)
+            {
+                replicaReplayTaskCts.Token.ThrowIfCancellationRequested();
+                var entryLength = storeWrapper.appendOnlyFile.HeaderSize;
+                var payloadLength = storeWrapper.appendOnlyFile.UnsafeGetLength(ptr);
+                if (payloadLength > 0)
+                {
+                    aofProcessor.ProcessAofRecordInternal(ptr + entryLength, payloadLength, true);
+                    entryLength += TsavoriteLog.UnsafeAlign(payloadLength);
+                }
+                else if (payloadLength < 0)
+                {
+                    if (!clusterProvider.serverOptions.EnableFastCommit)
+                    {
+                        throw new GarnetException("Received FastCommit request at replica AOF processor, but FastCommit is not enabled", clientResponse: false);
+                    }
+                    TsavoriteLogRecoveryInfo info = new();
+                    info.Initialize(new ReadOnlySpan<byte>(ptr + entryLength, -payloadLength));
+                    //TODO: Verify again that this does not write into the AOF
```

Comment:
do we still need this TODO?

Reply:
no, this TODO was completed. I removed it